shorebird_code_push/lib/src/shorebird_updater_io.dart

```diff
@@ -0,0 +1,160 @@
+import 'dart:async';
+import 'dart:ffi';
+import 'dart:isolate';
+
+import 'package:ffi/ffi.dart';
+import 'package:meta/meta.dart';
+import 'package:shorebird_code_push/src/generated/updater_bindings.g.dart';
+import 'package:shorebird_code_push/src/shorebird_updater.dart';
+import 'package:shorebird_code_push/src/updater.dart';
+
+@visibleForTesting
+
+/// Type definition for [Isolate.run].
+typedef IsolateRun = Future<R> Function<R>(
+  FutureOr<R> Function(), {
+  String? debugName,
+});
+
+/// {@template shorebird_updater_io}
+/// The Shorebird IO Updater.
+/// {@endtemplate}
+class ShorebirdUpdaterImpl implements ShorebirdUpdater {
+  /// {@macro shorebird_updater_io}
+  ShorebirdUpdaterImpl(this._updater, {IsolateRun? run})
+      : _run = run ?? Isolate.run {
+    try {
+      // If the Shorebird Engine is not available, this will throw an exception.
+      // FIXME: Run this in an isolate or refactor the updater to avoid risking
+      // a hang.
```

Comment:
```suggestion
      // FIXME: Run this in an isolate or refactor the updater to avoid risking
      // a hang.  If another thread is also calling into Shorebird at the same time
      // the underlying Rust code could block getting the config lock.
```